lib/pharos/salt_api.rb

```diff
@@ -0,0 +1,86 @@
+# frozen_string_literal: true
+require "net/http"
+
+module Pharos
+  # This class offers the integration between ruby and the Saltstack API.
+  module SaltApi
+    def self.included(base)
+      base.extend(ClassMethods)
+    end
+
+    attr_accessor :token
+
+    protected
+
+    # Perform request to endpoint, with method and provided data
+    def perform_request(endpoint:, method:, data: {})
+      self.class.perform_request endpoint: endpoint,
+                                 method:   method,
+                                 data:     data,
+                                 client:   self
+    end
+
+    # Defines class methods
+    module ClassMethods
+      # Performs a log in request and sets the @token instance variable on success at
+      # client object.
+      def login!(client:)
+        return client.token if client&.token
+        token = perform_request(endpoint: "/login",
+                                method:   "post",
+                                data:     auth_details)["X-Auth-Token"]
+        client&.token = token
+        token
+      end
+
+      # Returns the authentication details for the current environment.
+      def auth_details
+        {
+          username: ENV["PHAROS_SALT_USER"],
+          password: ENV["PHAROS_SALT_PASSWORD"],
+          eauth:    "pam"
+        }.freeze
+      end
+
+      # Returns salt hostname for the current environment.
+      def hostname
+        "#{ENV["PHAROS_SALT_HOST"]}:#{ENV["PHAROS_SALT_PORT"]}"
+      end
+
+      # Performs an HTTP request to the Salt API server. This methods already sets
+      # some headers like "X-Auth-Token" automatically.
+      #
+      # Returns the response object.
+      def perform_request(endpoint:, method:, data: {}, client: nil)
+        token = endpoint == "/login" ? nil : login!(client: client)
+
+        uri = URI.join("http://#{hostname}", endpoint)
+
+        req = Net::HTTP.const_get(method.capitalize).new(uri)
+        req["Accept"]       = "application/json; charset=utf-8"
+        req["Content-Type"] = "application/json; charset=utf-8"
+        req["X-Auth-Token"] = token if token
+        req.body = data.to_json unless data.blank?
+
+        opts = { use_ssl: false, open_timeout: 2 }
+        Net::HTTP.start(uri.hostname, uri.port, opts) { |http| http.request(req) }
+      end
+
+      # Perform an HTTP request to the Salt API server in a one-shot fashion.
+      #
+      # Returns the response object.
+      def perform_run_request(data: {})
+        uri = URI.join("http://#{hostname}", "run")
+
+        req = Net::HTTP::Post.new uri
+        req["Accept"]       = "application/json; charset=utf-8"
+        req["Content-Type"] = "application/json; charset=utf-8"
+
+        req.body = data.merge(auth_details).to_json
```

Comment:
No, the whole idea of the `/run` endpoint is to provide the authentication details within the data. It is similar but different in important ways to `perform_request`. I think we could end up with a mangled and hard to use method:

- It takes `data` only as argument.
- It will never save `token` or use it. Always send raw [auth details](https://docs.saltstack.com/en/latest/ref/netapi/all/salt.netapi.rest_cherrypy.html#post--run). Token is ignored, auth details are sent within the post request.
- It is always a `POST` request.

I'd keep them separate for the sake of readability.